### ISSUE TITLE
Enable jemalloc on 10.4, fix problems

### ISSUE THREAD

**Issue by GoogleCodeExporter** [classilla/tenfourfox#218]
```
Although I don't know if this is a contributor to issue 193, it sure seems like 
a prerequisite.
```

Original issue reported on code.google.com by `classi...@floodgap.com` on 13 Apr 2013 at 12:15
* Blocking: #193

**Comment by internetzel** [comment]
Just found out something quite important about the virtual memory management function madvise that is used in memory allocators like jemalloc to indicate the system whether memory can safely be paged out or not.
Since 10.3 there has been the option MADV_FREE defined in the mman.h system header which supposedly indicates the kernel that a virtual memory region is not in use any longer and hence can be paged out immediately - but it wasn't implemented in the kernel until 10.6 (see http://fxr.watson.org/fxr/trackident?i=MADV_FREE and select the use of MADV_FREE in different xnu versions).
Until 10.5 this option does simply nothing but return an error plus the waste of time of performing a syscall - as a consequence the kernel will have to trust it's own statistics about what memory to page out first, and building up meaningful statistics takes time - and can definitley cause a long running browser instance to crawl (see https://bugzilla.mozilla.org/show_bug.cgi?id=736074).
In order to help the kernel to quickly know what to page out first there's really only MADV_DONTNEED implemented up to including 10.5 .
I found this because in WebKit 601 TCMalloc was replaced by bmalloc (needing a C++11 compiler and runtime) and the browser started to hang in madvise calls quite soon after launch - I don't whether using MADV_DONTNEED instead of MADV_FREE is the solution but at least the syscall does succeed now (feels much better that way). Since the JavaScriptCore garbage collector now bases on bmalloc as well that thing has to work really well or whole WebKit becomes pretty much unusable.

**Comment by classilla** [comment]
Interesting. Something to consider for the future if I resurrect this again.

**Comment by internetzel** [comment]
While the unimplemented MADV_FREE is interesting I found that the cause for bmalloc spending much time in madvise() was in fact a non working custom mutex built around std::atomic_flag from libstdc++.
I'm pretty sure that rather simple mutex class is properly implemented, so maybe std::atomic_flag is not working correctly (the mutex is needed because freeing memory respectively marking memory as MADV_DONTNEED, for example for JavaScript garbage collection, happens in a separate thread).
By disabling bmalloc and using system malloc instead, the hangs are now gone.
And as WebKit uses memalign() I implemented a wrapper around malloc() that returns aligned adresses (wasting space of course). Since WebKit uses a seperate function for freeing aligned memory there's no problem in storing the original adress returned by malloc() in the extra space allocated for being able to do the alignment, so it shouldn't be leak memory.
But that non-working mutex seems alarming to me.
